### PR TITLE
Issue 2b - ec and rsa signing

### PR DESCRIPTION
**Pull request by ericeberry**

### CODE REVIEW

**Review by richard-garfinkel — APPROVED** [review]

File: reference/src/client/src/sa_common.c [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+            }
+
+            if (EC_POINT_oct2point(ec_group, ec_point, public_key, sizeof(public_key), NULL) != 1) {
+                ERROR("EC_POINT_oct2point failed");
+                break;
+            }
+
+            ec_key = EC_KEY_new_by_curve_name(type);
+            if (ec_key == NULL) {
+                ERROR("EC_KEY_new failed");
```
Comment (mhabrat) [inline reference/src/client/src/sa_common.c]: EC_KEY_new_by_curve_name?

File: reference/src/client/src/sa_common.c [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+                EVP_PKEY_free(evp_pkey);
+                break;
+            }
+        }
+#if OPENSSL_VERSION_NUMBER >= 0x10100000
+        else if (curve == SA_ELLIPTIC_CURVE_ED25519 || curve == SA_ELLIPTIC_CURVE_ED448 ||
+                 curve == SA_ELLIPTIC_CURVE_X25519 || curve == SA_ELLIPTIC_CURVE_X448) {
+            evp_pkey = EVP_PKEY_new_raw_public_key(type, NULL, in, in_length);
+            if (evp_pkey == NULL) {
+                ERROR("EVP_PKEY_new failed");
```
Comment (mhabrat) [inline reference/src/client/src/sa_common.c]: EVP_PKEY_new_raw_public_key?

File: reference/src/client/src/sa_engine_pkey.c [hunk 3]
```diff
@@ -0,0 +1,1069 @@
...
+        return UINT32_MAX;
+    }
+
+    return key_data->private_key;
+}
+
+static bool set_pkey_key_data(EVP_PKEY** evp_pkey, sa_key key) {
+    pkey_key_data* key_data = EVP_PKEY_get_ex_data(*evp_pkey, get_ex_data_index());
+    if (key_data == NULL) {
+        ERROR("OPENSSL_malloc failed");
```
Comment (mhabrat) [inline reference/src/client/src/sa_engine_pkey.c]: EVP_PKEY_get_ex_data?

File: reference/src/client/src/sa_engine_pkey.c [hunk 4]
```diff
@@ -0,0 +1,1069 @@
...
+    EVP_PKEY* evp_pkey = EVP_PKEY_CTX_get0_pkey(evp_pkey_ctx);
+    if (evp_pkey == NULL) {
+        ERROR("NULL evp_pkey");
+        return 0;
+    }
+
+    sa_key key = get_pkey_key_data(evp_pkey);
+    sa_header header;
+    if (sa_key_header(&header, key) != SA_STATUS_OK) {
+        ERROR("NULL evp_pkey");
```
Comment (mhabrat) [inline reference/src/client/src/sa_engine_pkey.c]: sa_key_header?

File: reference/src/client/src/sa_engine_pkey.c [hunk 5]
```diff
@@ -0,0 +1,1069 @@
...
+                if (app_data->pss_salt_length == RSA_PSS_SALTLEN_DIGEST) {
+                    salt_length = EVP_MD_size(app_data->evp_md);
+                } else if (app_data->pss_salt_length == RSA_PSS_SALTLEN_AUTO ||
+                           app_data->pss_salt_length == RSA_PSS_SALTLEN_MAX) {
+                    salt_length = EVP_PKEY_size(evp_pkey) - EVP_MD_size(app_data->evp_md) - 2;
+                    if ((EVP_PKEY_bits(evp_pkey) & 0x7) == 1)
+                        salt_length--;
+
+                    if (salt_length < 0) {
+                        ERROR("digest_algorithm unknown");
```
Comment (mhabrat) [inline reference/src/client/src/sa_engine_pkey.c]: salt_length?

File: reference/src/client/src/sa_engine_pkey.c [hunk 6]
```diff
@@ -0,0 +1,1069 @@
...
+        sa_key verify_key = get_pkey_key_data(evp_pkey);
+        verify_pkey = get_public_key(verify_key);
+        if (verify_pkey == NULL) {
+            ERROR("NULL verify_pkey");
+            break;
+        }
+
+        verify_pkey_ctx = EVP_PKEY_CTX_new(verify_pkey, NULL);
+        if (verify_pkey_ctx == NULL) {
+            ERROR("NULL verify_pkey_ctx");
```
Comment (mhabrat) [inline reference/src/client/src/sa_engine_pkey.c]: EVP_PKEY_CTX_new?

File: reference/src/client/src/sa_common.c (hunk 1 above)
Comment (ericeberry) [reply reference/src/client/src/sa_common.c]: Done

File: reference/src/client/src/sa_common.c (hunk 2 above)
Comment (ericeberry) [reply reference/src/client/src/sa_common.c]: Done

File: reference/src/client/src/sa_engine_pkey.c (hunk 3 above)
Comment (ericeberry) [reply reference/src/client/src/sa_engine_pkey.c]: Done

File: reference/src/client/src/sa_engine_pkey.c (hunk 4 above)
Comment (ericeberry) [reply reference/src/client/src/sa_engine_pkey.c]: Done

File: reference/src/client/src/sa_engine_pkey.c (hunk 5 above)
Comment (ericeberry) [reply reference/src/client/src/sa_engine_pkey.c]: Done

File: reference/src/client/src/sa_engine_pkey.c (hunk 6 above)
Comment (ericeberry) [reply reference/src/client/src/sa_engine_pkey.c]: Done

**Review by mhabrat — APPROVED** [review]

File: reference/src/client/src/sa_engine.c [hunk 7]
```diff
@@ -31,7 +31,14 @@ mtx_t engine_mutex;
 static once_flag flag = ONCE_FLAG_INIT;
 
 static void sa_engine_shutdown() {
```
Comment (kamaroyl) [inline reference/src/client/src/sa_engine.c:33]: Should the mtx_destroy happen first in order to de-initialize the mutex and make any subsequent attempts to lock the mutex fail?

Comment (ericeberry) [reply reference/src/client/src/sa_engine.c:33]: I'll switch it.

**Pull request merged**